app/src/main/java/com/example/javaopencommerce/order/OrderIntegrityValidator.java

```diff
@@ -0,0 +1,135 @@
+import static java.util.Optional.ofNullable;
+import static java.util.stream.Collectors.toMap;
+import static java.util.stream.Collectors.toUnmodifiableList;
+
+import com.example.javaopencommerce.item.ItemQueryRepository;
+import com.example.javaopencommerce.item.dtos.CardDto;
+import com.example.javaopencommerce.item.dtos.ItemDto;
+import com.example.javaopencommerce.item.dtos.ProductDto;
+import com.example.javaopencommerce.order.dtos.OrderDetailsDto;
+import com.example.javaopencommerce.order.exceptions.AddressNotProvidedValidationException;
+import com.example.javaopencommerce.order.exceptions.OrderStateValidationException;
+import com.example.javaopencommerce.order.exceptions.OrderValidationException;
+import com.example.javaopencommerce.order.exceptions.OrderValueNotMatchingValidationException;
+import com.example.javaopencommerce.order.exceptions.OutOfStockItemsValidationException;
+import io.smallrye.mutiny.Uni;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+
+class OrderIntegrityValidator {
+
+  private final ItemQueryRepository itemQueryRepository;
+  private List<OrderValidationException> orderInaccuracies;
+
+  OrderIntegrityValidator(ItemQueryRepository itemQueryRepository) {
+    this.itemQueryRepository = itemQueryRepository;
+  }
+
+  public Uni<Void> validateOrder(OrderDetailsDto order) {
+    orderInaccuracies = new ArrayList<>();
+    validateOrderBasics(order);
+    validateOrderState(order);
+    Uni<List<ItemDto>> currentItems = fetchCurrentItems(order.getCard().getProducts());
+    currentItems
+        .onItem().invoke(items -> validateOrderValue(items, order.getCard()))
+        .onItem().invoke(items -> validateStocks(items, order.getCard().getProducts()))
+        .await().indefinitely();
+
+    if (!orderInaccuracies.isEmpty()) {
+      throw new OrderValidationException(orderInaccuracies);
+    }
+
+    return Uni.createFrom().voidItem();
+  }
+
+  private Uni<List<ItemDto>> fetchCurrentItems(List<ProductDto> products) {
+    List<Long> itemIds = products.stream()
+        .filter(Objects::nonNull)
+        .map(ProductDto::getItem)
+        .filter(Objects::nonNull)
+        .map(ItemDto::getId)
+        .collect(toUnmodifiableList());
+    return itemQueryRepository.getItemsByIdList(itemIds);
+  }
+
+  private void validateOrderBasics(OrderDetailsDto order) {
+    if (order.getAddressId() == null) {
+      orderInaccuracies.add(new AddressNotProvidedValidationException());
+    }
+  }
+
+  private void validateOrderState(OrderDetailsDto order) {
+    if (order.getOrderStatus() == null || !order.getOrderStatus().equals("NEW")) {
+      orderInaccuracies.add(new OrderStateValidationException(
+          format("Wrong/Illegal order status: %s", order.getOrderStatus())));
+    }
+
+    if (order.getPaymentStatus() == null || !order.getPaymentStatus().equals("BEFORE_PAYMENT")) {
```

Comment:
As above:
```suggestion
    if ("BEFORE_PAYMENT".equals(order.getPaymentStatus())) {
```
Though BEFORE_PAYMENT is a enum string VALUE. It would be better to use that enum, and it's string value, to give **business context** to that comparison

Reply:
good idea, changed!